libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/location/CarLocationProvider.kt

```diff
@@ -33,16 +39,22 @@ internal class CarAppLocationImpl : CarAppLocation {
     }
 
     override fun onAttached(mapboxNavigation: MapboxNavigation) {
-        logAndroidAuto("CarAppLocationImpl onAttached")
         mapboxNavigation.registerLocationObserver(locationObserver)
     }
 
     override fun onDetached(mapboxNavigation: MapboxNavigation) {
-        logAndroidAuto("CarAppLocationImpl onDetached")
         mapboxNavigation.unregisterLocationObserver(locationObserver)
     }
 
-    override suspend fun validLocation(): Location? = withContext(Dispatchers.Unconfined) {
+    override fun registerLocationConsumer(locationConsumer: LocationConsumer) {
+        navigationLocationProvider.registerLocationConsumer(locationConsumer)
+    }
+
+    override fun unRegisterLocationConsumer(locationConsumer: LocationConsumer) {
+        navigationLocationProvider.unRegisterLocationConsumer(locationConsumer)
+    }
+
+    suspend fun validLocation(): Location? = withContext(Dispatchers.Unconfined) {
```

Comment:
This function should definitely be improved. No need to check `lastLocation` in a loop, because we can register a location consumer.

Reply:
But this is also many flowable operations work. Which operator supports "waiting for a location" without a loop?

```
public suspend fun <T> Flow<T>.firstOrNull(): T? {
    var result: T? = null
    collectWhile {
        result = it
        false
    }
    return result
}
```

Reply:
I can use firstOrNull but it still loops

Reply:
add a flow to store the lastest

Reply:
I mean avoid loops that check some condition every 100 milliseconds or anything like that. Current solution is fine, but I had `suspendCancelableCoroutine` in mind.